File: src/scenes/Scene.ts
import { Camera } from '../views/cameras/Camera';
// import { WebGLRenderer } from '../renderers/webgl/WebGLRenderer';
// import { WebGPURenderer } from '../renderers/webgpu/WebGPURenderer';
import { WebGLContext, WebGLContextOptions } from '../renderers/webgl/WebGLContext';
import { WebGPUContext, WebGPUContextOptions } from '../renderers/webgpu/WebGPUContext';

import type { Light } from '../lights/Light';
import type { RendererType } from '../renderers/Renderer';
import type { Renderable } from '../core/Model';

export class Scene {

    ready: boolean = false;

    private _canvas: HTMLCanvasElement | null = null;

    // 使用私有变量存储当前活动相机
    private _activeCamera: Camera | null = null;

    // private _renderer: WebGLRenderer  | WebGPURenderer | null = null;

    private _activeContext!: WebGLContext | WebGPUContext;
    private _webglContext: WebGLContext;
    private _webgpuContext: WebGPUContext;

    // 存储场景中的所有组件
    private components: any[] = [];

    // 新增：光照数组
    private lights: Light[] = [];

    constructor(canvas: string | HTMLCanvasElement, options: WebGLContextOptions | WebGPUContextOptions) {
        // 获取 canvas 元素
        if (canvas instanceof HTMLCanvasElement) {
            this._canvas = canvas;
        } else if (typeof canvas === 'string') {
            // 如果传入的是字符串，则尝试获取对应的 canvas 元素
            this._canvas = document.getElementById(canvas) as HTMLCanvasElement;
        }
        // 确保 canvas 元素存在
        if (!this._canvas) {
            throw new Error('Canvas element not found');
        }

        this._webglContext = new WebGLContext(this._canvas, options);
        this._webgpuContext = new WebGPUContext(this._canvas, options);
    }

    get activeCamera(): Camera | null {
        return this._activeCamera;
    }

    // 设置当前活动相机
    // 注意：如果需要更复杂的相机管理，可以考虑使用相机管理器
    set activeCamera(cam: Camera | null) {
        this._activeCamera = cam;
    }

    // get renderer(): WebGLRenderer | WebGPURenderer | null {
    //     // 返回当前场景的渲染器
    //     // 注意：这里假设场景只使用一个渲染器
    //     return this._renderer;
    // }

    // set renderer(renderer: WebGLRenderer | WebGPURenderer | null) {
    //     // 设置当前场景的渲染器
    //     this._renderer = renderer;
    // }   

    get context(): WebGLContext | WebGPUContext {
        // 返回当前场景的渲染器
        // 注意：这里假设场景只使用一个渲染器
        return this._activeContext;
    }

    set context(renderer: RendererType) {
        // 设置当前场景的渲染器
        if (renderer === 'webgl') {
            this._activeContext = this._webglContext;
        } else if (renderer === 'webgpu') {
            this._activeContext = this._webgpuContext;
        } else {
            throw new Error(`Invalid renderer type: ${renderer}, it must be one of "webgl" or "webgpu"`);
        }
    }   

    addComponent(component: any): void {
        this.components.push(component);
    }

    removeComponent(component: any): void {
        const index = this.components.indexOf(component);
        if (index !== -1) {
            this.components.splice(index, 1);
        }
    }

    getComponents(): Renderable[] {
        return this.components.slice();
    }

    addLight(light: Light): void {
        this.lights.push(light);
    }

    removeLight(light: Light): void {
        const idx = this.lights.indexOf(light);
        if (idx !== -1) this.lights.splice(idx, 1);
    }

    getLights(): Light[] {
        return this.lights.slice();
    }

    update(deltaTime: number): void {
        for (const comp of this.components) {
            if (typeof comp.update === 'function') {
                comp.update(deltaTime);
            }
        }
    }

    // render(context: WebGLRenderingContext): void {
    //     // 检查渲染器是否已经存在
    //     if (!this._renderer) {
    //         throw new Error('Renderer is not set for the scene');
    //     }

    //     //
    //     for (const comp of this.components) {
    //         if (typeof comp.render === 'function') {
    //             comp.render(context);
    //         }
    //     }
    // }
}

File: iEngine-Native/Engine/src/scenes/Scene.cpp
#include "iengine/scenes/Scene.h"
#include "iengine/core/Model.h"
#include "iengine/renderers/opengl/OpenGLContext.h"

#include <algorithm>
#include <stdexcept>
#include <iostream>

namespace iengine {
    Scene::Scene(std::shared_ptr<WindowInterface> window) : window_(window) {
        // 如果window为空，则不创建Context（适用于纯代码模式）
        if (!window_) {
            std::cout << "Scene created without window (headless mode)" << std::endl;
            return;
        }
        
        // 创建OpenGL Context
        openglContext_ = std::make_shared<OpenGLContext>(window_);
        openglContext_->init();
        
        // 默认使用OpenGL Context
        activeContext_ = openglContext_;
        
        std::cout << "Scene created with OpenGL context" << std::endl;
    }
    
    Scene::~Scene() {}
    
    void Scene::addComponent(std::shared_ptr<Model> component) {
        components_.push_back(component);
    }
    
    void Scene::removeComponent(std::shared_ptr<Model> component) {
        auto it = std::find(components_.begin(), components_.end(), component);
        if (it != components_.end()) {
            components_.erase(it);
        }
    }
    
    const std::vector<std::shared_ptr<Model>>& Scene::getComponents() const {
        return components_;
    }
    
    void Scene::addLight(std::shared_ptr<Light> light) {
        lights_.push_back(light);
    }
    
    void Scene::removeLight(std::shared_ptr<Light> light) {
        auto it = std::find(lights_.begin(), lights_.end(), light);
        if (it != lights_.end()) {
            lights_.erase(it);
        }
    }
    
    const std::vector<std::shared_ptr<Light>>& Scene::getLights() const {
        return lights_;
    }
    
    void Scene::update(float deltaTime) {
        for (auto& component : components_) {
            component->update(deltaTime);
        }
    }
    
    std::shared_ptr<Camera> Scene::getActiveCamera() const {
        return activeCamera_;
    }
    
    void Scene::setActiveCamera(std::shared_ptr<Camera> camera) {
        activeCamera_ = camera;
    }
    
    std::shared_ptr<Context> Scene::getContext() const {
        return activeContext_;
    }
    
    void Scene::setContextType(RendererType type) {
        switch (type) {
            case RendererType::OpenGL:
                if (!openglContext_) {
                    openglContext_ = std::make_shared<OpenGLContext>(window_);
                    openglContext_->init();
                }
                activeContext_ = openglContext_;
                std::cout << "Scene context set to OpenGL" << std::endl;
                break;
            default:
                throw std::runtime_error("Unsupported renderer type");
        }
    }
}